proto-lens-discrimination/src/Data/ProtoLens/Discrimination.hs

```diff
@@ -0,0 +1,136 @@
+-- | This module implements discriminators of protobufs generically.
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.ProtoLens.Discrimination
+    ( discFields
+    , discField
+    , discText
+    , discByteString
+    , discProtoMapAssocs
+    ) where
+
+import Data.Bits ((.|.), shift)
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Unsafe as B (unsafeIndex)
+import Data.Default (Default(def))
+import Data.Foldable (foldMap)
+import Data.Functor.Contravariant
+    ( Contravariant
+    , (>$<)
+    )
+import Data.Functor.Contravariant.Divisible
+    ( chosen
+    , conquered
+    , divided
+    , Decidable(choose)
+    )
+import Data.Map (Map)
+import Data.Word (Word8, Word16)
+import qualified Data.Map as M
+import qualified Data.Text as T
+import Lens.Family2 ((&), (.~), Lens', view)
+
+import Data.ProtoLens.Message
+    ( FieldDescriptor(FieldDescriptor)
+    , FieldTypeDescriptor(..)
+    , FieldAccessor(..)
+    )
+
+-- | Sort values according to a Foldable of field descriptors.
+discFields
```

Comment:
You'll probably also want to sort according to the `unknownFields` of the proto, added in #129: https://github.com/google/proto-lens/blob/master/proto-lens/src/Data/ProtoLens/Message.hs#L88

Up to you whether to do that in this PR or just leave it as a TODO.

Reply:
Added support for unknown fields.  Now 'sortingMessage' considers unknown fields and 'sortingKnownFields' does not.